python/cudf/cudf/core/indexing_utils.py

```diff
@@ -235,3 +270,344 @@ def parse_row_iloc_indexer(key: Any, n: int) -> IndexingSpec:
+
+    Parameters
+    ----------
+    needles
+        Labels to look for
+    haystack
+        Haystack to search in
+
+    Returns
+    -------
+    NumericalColumn
+        Integer gather map of locations needles were found in haystack
+
+    Raises
+    ------
+    KeyError
+        If not all needles were found in the haystack.
+        If needles cannot be converted to the dtype of haystack.
+
+    Notes
+    -----
+    This sorts the gather map so that the result comes back in the
+    order the needles were specified (and are found in the haystack).
+    """
+    # Pre-process to match dtypes
+    needle_kind = needles.dtype.kind
+    haystack_kind = haystack.dtype.kind
+    if haystack_kind == "O":
+        try:
+            needles = needles.astype(haystack.dtype)
+        except ValueError:
+            # Pandas raise KeyError here
+            raise KeyError("Dtype mismatch in label lookup")
+    elif needle_kind == haystack_kind or {
+        haystack_kind,
+        needle_kind,
+    }.issubset({"i", "u", "f"}):
+        needles = needles.astype(haystack.dtype)
+    elif needles.dtype != haystack.dtype:
+        # Pandas raise KeyError here
+        raise KeyError("Dtype mismatch in label lookup")
+    # Can't always do an inner join because then we can't check if we
+    # had missing keys (can't check the length because the entries in
+    # the needle might appear multiple times in the haystack).
+
+    left_rows, right_rows = plc.join.left_join(
+        plc.Table([needles.to_pylibcudf(mode="read")]),
+        plc.Table([haystack.to_pylibcudf(mode="read")]),
+        plc.types.NullEquality.EQUAL,
+    )
+    rgather = (
+        type(haystack)
+        .from_pylibcudf(right_rows)
+        ._with_type_metadata(haystack.dtype)
+    )
+    (right_order_plc,) = copying.gather(
+        [as_column(range(len(haystack)), dtype=np.dtype(np.int32))],
+        rgather,  # type: ignore[arg-type]
+        nullify=True,
+    )
+    right_order = type(haystack).from_pylibcudf(right_order_plc)
+    if right_order.null_count > 0:
+        raise KeyError("Not all keys in index")
+    lgather = (
+        type(needles)
+        .from_pylibcudf(left_rows)
+        ._with_type_metadata(needles.dtype)
+    )
+    (left_order_plc,) = copying.gather(
+        [as_column(range(len(needles)), dtype=np.dtype(np.int32))],
+        lgather,  # type: ignore[arg-type]
+        nullify=False,
+    )
+    left_order = type(needles).from_pylibcudf(left_order_plc)
+    (rgather_plc,) = sorting.sort_by_key(
+        [rgather],
+        [left_order, right_order],
+        [True, True],
+        ["last", "last"],
+        stable=True,
+    )
+    rgather = type(rgather).from_pylibcudf(rgather_plc)
+    return GatherMap.from_column_unchecked(
+        rgather,  # type: ignore[arg-type]
+        len(haystack),
+        nullify=False,
+    )
+
+
+def find_label_range_or_mask(
+    key: slice, index: Index
+) -> EmptyIndexer | MapIndexer | MaskIndexer | SliceIndexer:
```

Comment:
*question*: This return type is broader than the values we return (only `EmptyIndexer` or `SliceIndexer`), can we tighten it here?

Reply:
Yup, done